fw/http.c

```diff
@@ -3430,57 +3460,127 @@ static int
 tfw_h1_adjust_req(TfwHttpReq *req)
 {
 	int r;
-	unsigned int n_to_strip = 0;
 	TfwHttpMsg *hm = (TfwHttpMsg *)req;
+	TfwStr *pos, *end;
+	const TfwStr *meth;
+	static const DEFINE_TFW_STR(meth_get, "GET");
+	static const DEFINE_TFW_STR(slash, "/");
+	static const DEFINE_TFW_STR(sp, " ");
+	static const DEFINE_TFW_STR(crlf, S_CRLF);
+	static const DEFINE_TFW_STR(ver, " " S_VERSION11 S_CRLF);
+	const TfwHdrMods *h_mods = tfw_vhost_get_hdr_mods(req->location,
+							  req->vhost,
+							  TFW_VHOST_HDRMOD_REQ);
 
-	n_to_strip = !!test_bit(TFW_HTTP_B_NEED_STRIP_LEADING_CR, req->flags) +
-		     !!test_bit(TFW_HTTP_B_NEED_STRIP_LEADING_LF, req->flags);
-	if (unlikely(n_to_strip)) {
-		r =  ss_skb_list_chop_head_tail(&hm->msg.skb_head, n_to_strip, 0);
-		if (r)
-			return r;
-	}
+	req->cleanup = tfw_pool_alloc(hm->pool, sizeof(TfwHttpMsgCleanup));
+	if (unlikely(!req->cleanup))
+		return -ENOMEM;
+	memset(req->cleanup, 0, sizeof(TfwHttpMsgCleanup));
 
-	if (test_bit(TFW_HTTP_B_PURGE_GET, req->flags)) {
-		r = tfw_h1_rewrite_purge_to_get(&hm->msg.skb_head);
-		if (r)
-			return r;
-	}
+	tfw_msg_transform_setup(&req->iter, req->msg.skb_head);
+	r = tfw_http_msg_cutoff_headers(hm, req->cleanup);
+	if (unlikely(r))
+		goto clean;
 
-	r = tfw_http_sess_req_process(req);
-	if (r)
-		return r;
+	if (test_bit(TFW_HTTP_B_PURGE_GET, req->flags))
+		/* Rewrite PURGE to GET */
+		meth = &meth_get;
+	else
+		meth = &hm->h_tbl->tbl[TFW_HTTP_METHOD];
+
+	r = tfw_http_msg_expand_from_pool(hm, meth);
+	if (unlikely(r))
+		goto clean;
+
+	r = tfw_http_msg_expand_from_pool(hm, &sp);
+	if (unlikely(r))
+		goto clean;
+
+	if (TFW_STR_EMPTY(&req->uri_path))
+		r = tfw_http_msg_expand_from_pool(hm, &slash);
+	else
+		r = tfw_http_msg_expand_from_pool(hm, &req->uri_path);
+
+	if (unlikely(r))
+		goto clean;
+
+	r = tfw_http_msg_expand_from_pool(hm, &ver);
+	if (unlikely(r))
+		goto clean;
+
+	r = tfw_http_add_hdr_host(req);
+	if (unlikely(r))
+		goto clean;
+
+	FOR_EACH_HDR_FIELD_FROM(pos, end, hm, TFW_HTTP_HDR_CONTENT_LENGTH) {
+		int hid = pos - hm->h_tbl->tbl;
+		TfwStr *dup, *dup_end, *hdr = pos;
+
+		/* Skip hop-by-hop headers. */
+		if (TFW_STR_EMPTY(hdr) || hdr->flags & TFW_STR_HBH_HDR
+		    || hid == TFW_HTTP_HDR_X_FORWARDED_FOR) {
+			continue;
+		}
+
+		if (TFW_STR_DUP(hdr))
+			hdr = TFW_STR_CHUNK(hdr, 0);
+
+		if (hid == TFW_HTTP_HDR_CONTENT_TYPE
+		    && req->method == TFW_HTTP_METH_POST
+		    && test_bit(TFW_HTTP_B_CT_MULTIPART, req->flags)
+		    && tfw_http_should_validate_post_req(req))
+		{
+			r = tfw_http_recreate_content_type_multipart_hdr(req);
+			if (unlikely(r))
+				return r;
+			continue;
+		}
+
+		if (tfw_http_hdr_sub(hid, hdr, h_mods))
+			continue;
```

Comment:
Headers substitution splitted into two parts: first `tfw_http_hdr_sub()` where we just skip header and don't add it to message. Second `tfw_h1_add_loc_hdrs()` where we adding all headers. We don't do it in `tfw_http_hdr_sub()` because in this case we need add additional if condition into `tfw_h1_add_loc_hdrs()` to skip headers if `append == 0` - resp/req_header_set case.

Name `tfw_http_hdr_sub()` looks little bit confusing, I will rename it.